keps/sig-api-machinery/365-paginated-lists/README.MD

```diff
@@ -152,6 +248,86 @@ malicious users cannot trick the server into serving results outside of its rang
+  support index by `spec.nodeName`.
+  There are two main things we can do here:
+  - instead of storing an index in map {value}->{list of keys of objects whose index matches value}
+    switch to also use a BTree for the index. This would be the most complete implementation although
+    the drawback is that with the copy-on-write semantics, the update of the list means effectively
+    a need to copy the list and inserting the modified one. This can become expensive if those lists
+    become longer.
+  - drop support indexing for continuations. The index is just an optimization so we don't need it
+    for correctness (on the cost of less efficient processing). However, given that we only support
+    indexing by `spec.nodeName` and the default LIMIT is set to 500, in huge majority of cases
+    (in particular Kubelet listing its pods before watching) the whole response will get returned
+    in the single call. As a result, we may consider not cloning the index itself and in case
+    continuation request will be needed, just go over all objects without the index.
+
+  Given that the index is just an optimization and the need for it is not clear for now, we will
+  just start with the second option above (dropping support for indexes for continuations) and
+  will reconsider adding that once we observe the need for it.
+
+2. The HA kube-apiserver. The fact that the original request was processed by kube-apiserver A
+  doesn't mean that the continuation request will not be send to kube-apiserver B. In the following
+  sketch of the algorithm, kube-apiaserver B will not have the cloned state from the given RV
+  though.
+  Again we have two main approaches here:
+  - for every transaction processed by watchcache, we will be making a copy-on-write copy of the
+    current state. That means that transaction will not only contain the data about what was
+    changed, but also a point to the state from that point (copy-on-write BTree).
+  - ignore that usecase for now, and just delegate the continuation request to etcd if the
+    copy-on-write state for a resource version specified by continuation doesn't exist in memory
+
+  Similarly as above, processing continuation in watchcache is just an optimization and we don't
+  strictly need it. However, the first option is simple to implement to we will implement it
+  and use it unless a significant cpu/memory overhead will be shown by the benchmarks.
+
+3. The length of `out-of-history` window. Currently in watchcache we're storing transactions from
+  the last ~75 seconds (we're clearing transactions older than that when a new transaction is
+  being processed). This means that continuation will be served from watchcache only for ~75
+  seconds. However, in etcd have 2.5 to 5 minutes of history. As a result, we may consider
+  extending the length of the transaction log to match the minimum 2.5 minutes from above.
```

Comment:
Do you have any projections on how this change would affect apiserver memory usage?

Reply:
I expect in many cases (e.g. few objects with high churn) this would lead to significant increase in RAM consumption.